src/lib/QueryType.hs

```diff
@@ -0,0 +1,849 @@
+  computeAbsEffects, declNestEffects,
+  caseAltsBinderTys, depPairLeftTy, extendEffect,
+  getAppType, getTabAppType, getMethodType, getBaseMonoidType, getReferentTy,
+  getMethodIndex,
+  instantiateDataDef, instantiateDepPairTy, instantiatePi, instantiateTabPi,
+  litType, lamExprTy,
+  numNaryPiArgs, naryLamExprType,
+  oneEffect, projectLength, sourceNameType, typeAsBinderNest, typeBinOp,
+  isSingletonType, singletonTypeVal,
+  ) where
+
+import Control.Monad
+import Data.Foldable (toList)
+import Data.List (elemIndex)
+import qualified Data.List.NonEmpty as NE
+import qualified Data.Map.Strict as M
+import Data.Maybe (isJust)
+import qualified Data.Set        as S
+
+import CheapReduction (cheapNormalize)
+import Err
+import LabeledItems
+import Name
+import Syntax
+import Util
+import PPrint ()
+
+-- === Main API for querying types ===
+
+getTypeSubst :: (SubstReader AtomSubstVal m
+                , EnvReader2 m, HasType e)
+             => e i -> m i o (Type o)
+getTypeSubst e = do
+  subst <- getSubst
+  liftTypeQueryM subst $ getTypeE e
+{-# INLINE getTypeSubst #-}
+
+getType :: (EnvReader m, HasType e) => e n -> m n (Type n)
+getType e = liftTypeQueryM idSubst $ getTypeE e
+{-# INLINE getType #-}
+
+-- === Querying effects ===
+
+getEffects :: (EnvReader m, HasEffectsE e) => e n -> m n (EffectRow n)
+getEffects e = liftTypeQueryM idSubst $ getEffectsImpl e
+{-# INLINE getEffects #-}
+
+getEffectsSubst :: (EnvReader2 m, SubstReader AtomSubstVal m, HasEffectsE e)
+              => e i -> m i o (EffectRow o)
+getEffectsSubst e = do
+  subst <- getSubst
+  liftTypeQueryM subst $ getEffectsImpl e
+{-# INLINE getEffectsSubst #-}
+
+-- === Exposed helpers for querying types and effects ===
+
+caseAltsBinderTys :: (Fallible1 m, EnvReader m)
+                  => Type n -> m n [EmptyAbs (Nest Binder) n]
+caseAltsBinderTys ty = case ty of
+  TypeCon _ defName params -> do
+    def <- lookupDataDef defName
+    cons <- applyDataDefParams def params
+    return [bs | DataConDef _ bs <- cons]
+  VariantTy (NoExt types) -> do
+    mapM typeAsBinderNest $ toList types
+  VariantTy _ -> fail "Can't pattern-match partially-known variants"
+  SumTy cases -> mapM typeAsBinderNest cases
+  _ -> fail $ "Case analysis only supported on ADTs and variants, not on " ++ pprint ty
+
+depPairLeftTy :: DepPairType n -> Type n
+depPairLeftTy (DepPairType (_:>ty) _) = ty
+{-# INLINE depPairLeftTy #-}
+
+extendEffect :: Effect n -> EffectRow n -> EffectRow n
+extendEffect eff (EffectRow effs t) = EffectRow (S.insert eff effs) t
+
+getAppType :: EnvReader m => Type n -> [Atom n] -> m n (Type n)
+getAppType f xs = liftTypeQueryM idSubst $ typeApp f xs
+{-# INLINE getAppType #-}
+
+getTabAppType :: EnvReader m => Type n -> [Atom n] -> m n (Type n)
+getTabAppType f xs = case nonEmpty xs of
+  Nothing -> getType f
+  Just xs' -> liftTypeQueryM idSubst $ typeTabApp f xs'
+{-# INLINE getTabAppType #-}
+
+getMethodType :: EnvReader m => ClassName n -> Int -> m n (Type n)
```

Comment:
This is exported, but I can't find any uses. Perhaps remove it form the export list to allow GHC to specialize/inline this internally to this file?

Reply:
Unexported.

Reply:
Never mind, it's actually used in CheckType.